Source/Core/InputCommon/ControllerInterface/evdev/evdev.cpp

```diff
@@ -186,91 +227,105 @@ void Shutdown()
+  node.device = dev;
 
   // Buttons (and keyboard keys)
-  int num_buttons = 0;
-  for (int key = 0; key < KEY_MAX; key++)
+  for (int key = 0; key != KEY_CNT; key++)
   {
-    if (libevdev_has_event_code(m_dev, EV_KEY, key))
-      AddInput(new Button(num_buttons++, key, m_dev));
+    if (libevdev_has_event_code(dev, EV_KEY, key))
+      AddInput(new Button(node.button_count++, key, node));
   }
 
   // Absolute axis (thumbsticks)
-  int num_axis = 0;
-  for (int axis = 0; axis < 0x100; axis++)
+  for (int axis = 0; axis < ABS_CNT; axis++)
   {
-    if (libevdev_has_event_code(m_dev, EV_ABS, axis))
+    if (libevdev_has_event_code(dev, EV_ABS, axis))
     {
-      AddAnalogInputs(new Axis(num_axis, axis, false, m_dev),
-                      new Axis(num_axis, axis, true, m_dev));
-      ++num_axis;
+      AddAnalogInputs(new Axis(node.axis_count, axis, false, node),
+                      new Axis(node.axis_count, axis, true, node));
+      ++node.axis_count;
     }
   }
 
+  u32 button_count = 0;
+  u32 axis_count = 0;
+  for (auto& [devnode, node] : m_nodes)
+  {
+    node.button_start = button_count;
+    node.axis_start = axis_count;
+
+    button_count += node.button_count;
+    axis_count += node.axis_count;
+  }
+
   // Disable autocenter
-  if (libevdev_has_event_code(m_dev, EV_FF, FF_AUTOCENTER))
+  if (libevdev_has_event_code(dev, EV_FF, FF_AUTOCENTER))
   {
     input_event ie = {};
     ie.type = EV_FF;
     ie.code = FF_AUTOCENTER;
     ie.value = 0;
 
-    static_cast<void>(write(m_fd, &ie, sizeof(ie)));
+    static_cast<void>(write(fd, &ie, sizeof(ie)));
   }
 
+  // FYI: Force names will collide if two nodes with the same "uniq" name support effects.
+  // I don't think that is going to be a problem.
+
   // Constant FF effect
-  if (libevdev_has_event_code(m_dev, EV_FF, FF_CONSTANT))
+  if (libevdev_has_event_code(dev, EV_FF, FF_CONSTANT))
   {
-    AddOutput(new ConstantEffect(m_fd));
+    AddOutput(new ConstantEffect(fd));
   }
 
   // Periodic FF effects
-  if (libevdev_has_event_code(m_dev, EV_FF, FF_PERIODIC))
+  if (libevdev_has_event_code(dev, EV_FF, FF_PERIODIC))
   {
     for (auto wave : {FF_SINE, FF_SQUARE, FF_TRIANGLE, FF_SAW_UP, FF_SAW_DOWN})
     {
-      if (libevdev_has_event_code(m_dev, EV_FF, wave))
-        AddOutput(new PeriodicEffect(m_fd, wave));
+      if (libevdev_has_event_code(dev, EV_FF, wave))
+        AddOutput(new PeriodicEffect(fd, wave));
     }
   }
 
   // Rumble (i.e. Left/Right) (i.e. Strong/Weak) effect
-  if (libevdev_has_event_code(m_dev, EV_FF, FF_RUMBLE))
+  if (libevdev_has_event_code(dev, EV_FF, FF_RUMBLE))
   {
-    AddOutput(new RumbleEffect(m_fd, RumbleEffect::Motor::Strong));
-    AddOutput(new RumbleEffect(m_fd, RumbleEffect::Motor::Weak));
+    AddOutput(new RumbleEffect(fd, RumbleEffect::Motor::Strong));
+    AddOutput(new RumbleEffect(fd, RumbleEffect::Motor::Weak));
   }
 
   // TODO: Add leds as output devices
 
   // Was there some reasoning behind these numbers?
```

Comment:
Highly probable that the reasoning is: You need two axes for a joystick, and eight buttons for a NES controller.